GX-TimeFrequency/src/minimax_grids.F90

```diff
@@ -0,0 +1,499 @@
+            0.0_dp, vec_UT_psi, num_x_nodes)
+
+       ! 3) (V*Sigma^-1) * (U^T*psi)
+       call dgemm('N', 'N', num_points, 1, num_x_nodes, 1.0_dp, mat_VT_s, num_points, vec_UT_psi, &
+            num_x_nodes, 0.0_dp, weights_work, num_points)
+
+       weights(i_point, :) = weights_work(:)
+
+       ! calculate the maximum error of the fitting
+       call calculate_max_error(num_points, tau_points, omega_points, weights_work, num_x_nodes, x_mu, &
+            psi, current_point, max_error, transformation_type)
+    end do ! i_point
+
+    deallocate (x_mu, psi, mat_A, weights_work, vec_S, mat_U, mat_VT, work, iwork, mat_VT_s, vec_UT_psi)
+
+  end subroutine get_transformation_weights
+
+  !> \brief Calculate the auxiliary matrix for cosine/sin transformation for tau to omega or viceversa
+  !! @param[in] num_points: Number of mesh points.
+  !! @param[in] tau_points: imaginary time grid points
+  !! @param[in] omega_points: imaginary frequency grid points
+  !! @param[in] num_x_nodes: Number of node in the interval [e_min,e_max]
+  !! @param[in] x_mu : Transition energy (nodes in the interval [e_min,e_max])
+  !! @param[inout] psi: corresponding auxiliary function (see transformation type definition)
+  !! @param[inout] mat_A: auxiliary matrix (see transformation type definition)  
+  !! @param[in] i_point: pointer for the current grid point
+  !! @param[inout] current_point:  current grid point ether omega(i_point) or tau_(i_point)
+  !! @param[in] transformation type :
+  !!        (1) the cosine transform cos(it) -> cos(iw): psi(omega,x), mat_A = cos(omega*tau)*psi(tau,x) 
+  !!        (2) the cosine transform cos(iw) -> cos(it): psi(tau,x)  , mat_A = cos(omega*tau)*psi(omega,x)
+  !!        (3) the sine transform   sin(it) -> sin(iw): psi(omega,x), mat_A = sin(omega*tau)*psi(tau,x)
+  subroutine calculate_psi_and_mat_A(num_points, tau_points, omega_points, num_x_nodes, x_mu, psi, &
+       mat_A, i_point, current_point, transformation_type)
+
+    integer, intent(in)                                :: num_points, num_x_nodes, i_point
+    real(kind=dp), allocatable, dimension(:), &
+         intent(in)                                    :: tau_points, omega_points, x_mu
+    real(kind=dp), allocatable, dimension(:), &          
+         intent(inout)                                 :: psi         
+    real(kind=dp), allocatable, dimension(:, :), &
+         intent(inout)                                 :: mat_A
+    real(kind=dp), intent(inout)                       :: current_point
+    integer, intent(in)                                :: transformation_type
+
+    ! Internal variables
+    integer                                            :: i_node, j_point
+    real(kind=dp)                                      :: tau, omega
+
+    ! Begin work
+
+    ! the cosine transform cos(it) -> cos(iw)
+    if (transformation_type == cosine_tw) then
```

Comment:
Well, I'm disappointed you disregarded my suggestion on removing duplicate code. For completeness, it would have looked like:

```fortran
   if (transformation_type == cosine_tw) then
      call fill_psi_and_A(omega_points(i_point), tau_points, x_mu, lorentz, cos_exp, psi, A_mat)

   else if (transformation_type == cosine_wt) then
      call fill_psi_and_A(tau_points(i_point), omega_points, x_mu, ADD_ME, ADD_ME, psi, A_mat)

   else if (transformation_type == sine_tw) then
      call fill_psi_and_A(omega_points(i_point), omega_points, x_mu, ADD_ME, ADD_ME, psi, A_mat)

   endif
```

with the routines defined as:

```fortran
  ! Give the routine a better name
   subroutine fill_psi_and_A(current_point, transform_points, x_mu, psi_func, A_func, psi, A_mat)
     ! Add declarations

     ! Write abstract interface called psi_func_def, as I showed in prior example
     procedure(psi_func_def) ::  psi_func
     procedure(A_func_def) ::  A_func

    ! If you wanted to improve this some more, make (for example) lorentz elemental
    ! Then you can scrap the loop in fill_psi_and_A and just have
    !psi = lorentz((x_mu, omega)
    do i = 1, size(x_mu)
       psi(i) = psi_func(x_mu(i), omega)
    end do

    do j = 1, num_points
       point = transform_points(j_point)
       do i = 1, num_x_nodes
          mat_A(i, j) = A_func(x_mu(i), omega, point)
       end do
    end do

  end subroutine fill_psi_and_A
```

and inject:

```fortran
   ! 2x/(x^2+omega_k^2)
   function lorentz(x_mu, omega)
     lorentz = 2.0_dp*x_mu/((x_mu)**2 + omega**2)
   end function lorentz


   ! cos(omega_k * tau) psi(tau,x)
   function cos_exp(x_mu, omega, tau)
     cos_exp = cos(omega*tau)*exp(-x_mu*tau)
   end function cos_exp
```

And the use of elemental would have reduced the code further.  Additionally, I would have split the construction of psi and A.

In the future, if have duplicate code, it's a code smell - optimise it. Was less than 1 hour of work